crates/hir_def/src/type_ref.rs

```diff
@@ -51,6 +51,25 @@ impl Rawness {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub enum TraitRef {
+    Path(Path),
+    Error,
```

Comment:
This gives me a pause.  Is using an explicit error node the best solution here? It feels like it, but I am not sure.

Reply:
We could just return an Option here, then the question is as florian stated earlier
> I guess the question is whether we're fine treating impl {syntax error} for Type the same as impl Type... and we probably are 🤔

Reply:
I think option would be worse. While in the AST we don't distinguish between optional and missing nodes, I think in the hir layer we should use some form of synthtic placeholder value, and only use Option for truly Optional things. The question is, what placeholder to use? There are several choices:

* If a required part is missing, just remove the node from hir altogether. This prevents analysis working in bodies, and we probably don't want this.
* Missclassify  the hir node: treat `impl Foo for {}` as `impl Foo {}`. That probably is the simplest solution, and the one we should choose, but it feels like a hack, as we are doing a guaranteed wrong thing
* Add an explicit missing variant. That's lossless, but all the call-sites now need to care about handling this variant.
  * Add `Result<T, Missing>` as a generalized patter for this.
* Add an implicit missing variant. That is, if there's no `Path`, manufacture a `Path` out of thin air that is known to not resolve anywhere. I think that's the cleanest approach overall. If you squint, `impl Foo for `  looks kind of like it impls the ` ` trait -- the trait designated by an empty path, which is known to not resolve to anything.

Reply:
Oh, I think I misunderstood the `Option` here. I think ` treating impl {syntax error} for Type the same as impl Type` is probably the right soulution right now, but we should keep the "implicit missing" thing in mind.  Which is fun -- I remember being annoyed at us having a `Name::missing`, variant, but now that I spelled that the above out, I think `missing` is the right, mathematically best approach, as long as we guarantee that each instance of `missing` is unique, such that a missing name can only equal itself.

Reply:
Ye I like that plan, changed it to `Option` with a FIXME for now